ember-core/src/main/scala/org/http4s/ember/core/Encoder.scala

```diff
@@ -0,0 +1,49 @@
+package org.http4s.ember.core
+
+import cats.effect._
+import fs2._
+import org.http4s._
+import cats._
+import cats.implicits._
+import Shared._
+
+private[ember] object Encoder {
+
+  def respToBytes[F[_]: Sync](resp: Response[F]): Stream[F, Byte] = {
+    val headerStrings: List[String] =
+      resp.headers.toList.map(h => h.name.show + ": " + h.value).toList
+
+    val initSection = Stream(show"${resp.httpVersion} ${resp.status}") ++ Stream.emits(
+      headerStrings)
+
+    val body = Alternative[Option]
+      .guard(resp.isChunked)
+      .fold(resp.body)(_ => resp.body.through(ChunkedEncoding.encode[F]))
+
+    initSection.covary[F].intersperse("\r\n").through(text.utf8Encode) ++
+      Stream.chunk(Chunk.ByteVectorChunk(`\r\n\r\n`)) ++
+      body
+  }
+
+  def reqToBytes[F[_]: Sync](req: Request[F]): Stream[F, Byte] = {
+    // Request-Line   = Method SP Request-URI SP HTTP-Version CRLF
+    val requestLine = show"${req.method} ${req.uri.renderString} ${req.httpVersion}"
+
+    val finalHeaders =
+      req.uri.authority
+        .fold(Headers.of())(auth => Headers.of(Header("Host", auth.renderString))) ++ req.headers
+
+    val headerStrings: List[String] =
+      finalHeaders.toList.map(h => h.name.show + ": " + h.value).toList
+
+    val initSection = Stream(requestLine) ++ Stream.emits(headerStrings)
+
+    val body = Alternative[Option]
+      .guard(req.isChunked)
+      .fold(req.body)(_ => req.body.through(ChunkedEncoding.encode[F]))
+
+    initSection.covary[F].intersperse("\r\n").through(text.utf8Encode) ++
```

Comment:
I have a question about this encoding/decoding. 

Now, the goal of this encoding is to generate a stream of bytes (a sequence of vectors of bytes) that are to be sent through the network, which involves an I/O operation and an Operating System call, which may be costly. If so, one may seek to reduce number of I/O calls, by _buffering_ as many of those as possible. In this context, would it make sense to "aggregate" as many of these strings or byte-vectors as possible into a single ByteVector page, which may be sent to the O/S in a _single_ step? 

At present, I am not sure if each element in the stream would involve a different OS call.

Reply:
The fastest thing is to build one String for the entire header and encode it in one pass.  This outperforms fs2 Pipes and NIO's incremental charset encoder (which I believe the pipes use).

Reply:
A recommended way to build a long strings, made of concatenating many small strings, is by using a `StringBuilder`. Is there also a single-pass encoder for that? Or a `ByteVector`-builder? 

IMHO, a mutable object that is local to a method is an acceptable impurity.

Reply:
Encoding incrementally to `ByteVector` is slow.  I've created `ByteVectorWriter` a few times and thrown it away each time because it never wins.  

There's no way that I'm aware of to encode directly from a `StringBuilder`, though `toString()` is cheap.